src/Confluent.Kafka/AdminClient.cs

```diff
@@ -0,0 +1,374 @@
+        public event EventHandler<Error> OnError;
+
+        /// <summary>
+        ///     Raised on librdkafka statistics events. JSON formatted
+        ///     string as defined here: https://github.com/edenhill/librdkafka/wiki/Statistics
+        /// </summary>
+        /// <remarks>
+        ///     You can enable statistics and set the statistics interval
+        ///     using the statistics.interval.ms configuration parameter
+        ///     (disabled by default).
+        ///
+        ///     Called on the Producer poll thread.
+        /// </remarks>
+        public event EventHandler<string> OnStatistics;
+
+        /// <summary>
+        ///     Raised when there is information that should be logged.
+        /// </summary>
+        /// <remarks>
+        ///     By default not many log messages are generated.
+        /// 
+        ///     You can specify one or more debug contexts using the 'debug'
+        ///     configuration property and a log level using the 'log_level'
+        ///     configuration property to enable more verbose logging,
+        ///     however you shouldn't typically need to do this.
+        ///
+        ///     Warning: Log handlers are called spontaneously from internal librdkafka 
+        ///     threads and the application must not call any Confluent.Kafka APIs from 
+        ///     within a log handler or perform any prolonged operations.
+        /// </remarks>
+        public event EventHandler<LogMessage> OnLog;
+        
+        /// <summary>
+        ///     Gets the name of this producer instance.
+        ///     Contains (but is not equal to) the client.id configuration parameter.
+        /// </summary>
+        /// <remarks>
+        ///     This name will be unique across all producer instances
+        ///     in a given application which allows log messages to be
+        ///     associated with the corresponding instance.
+        /// </remarks>
+        public string Name
+            => kafkaHandle.Name;
+        
+        /// <summary>
+        ///     Releases all resources used by this AdminClient.
+        /// </summary>
+        public void Dispose()
+        {
+            if (!this.manualPoll)
+            {
+                // Note: It's necessary to wait on callbackTask before disposing kafkaHandle.
+                // TODO: If called in a finalizer, can callbackTask be potentially null?
+                callbackCts.Cancel();
+                callbackTask.Wait();
+            }
+            kafkaHandle.Dispose();
+        }
+
+        /// <summary>
+        ///     Get information pertaining to all groups in the Kafka cluster (blocking)
+        ///
+        ///     [UNSTABLE-API] - The API associated with this functionality is subject to change.
+        /// </summary>
+        /// <param name="timeout">
+        ///     The maximum period of time the call may block.
+        /// </param>
+        public List<GroupInfo> ListGroups(TimeSpan timeout)
```

Comment:
The reason I marked this as unstable originally on Producer and Consumer is I really want the client types to be very focussed on their purpose. Does this method make sense on a thing called producer? No IMO. 

If we start exposing a new `Handle` type on clients, and allow other clients to be constructed from this (and note: clients use the handle, not are a handle), we can get both a clean API and avoid the need for unnecessary creation of connections to brokers. This is not possible with Producers and Consumers, because we need to specify the type at construction time, but it is possible to construct and AdminClient in this way.

But should this method be on AdminClient? Actually no as well - AdminClient should probably be focussed on what is provided by the Kafka Administrative API.

Should this method be on Consumer? Probably not IMO as well - the group protocol is intended as a general thing (not just for consumer groups), and our implementation does not yet provide anything specific to it's use for consumer groups. If we started interpreting the MemberMetadata and MemberAssignment byte arrays as consumer group data, we should definitely put that functionality on the Consumer though.

Perhaps ListGroups and GetMetadata should go on this new publicly exposed Handle class. Maybe all things that don't belong on Producer, Consumer or AdminClient make sense on Handle.

Actually, very much liking that idea ATM.

Handle would be very much like SafeKafkaHandle, except it would only expose these selected methods. It would probably wrap SafeKafkaHandle (which would be an internal property).

Reply:
What seems important from me (from a user perspective) is to have interfaces IProducer, IConsumer, IHandle, IAdmin. IProducer and IConsumer are linked to librdkafka reserved method for given type, but everything else is accessible from the Handle, which would implement IHandle and IAdmin (Handle being internal)

One difficulty is how to decide if some api are from IHandle or from IAdmin (ie, I don't see why 
describeTopics woul be an admin api, and getMetadata an handle one...) We can just stick to java, it's an open question :)

I'm ok with the Handle class. Just want to be able to access IAdmin from Producer (either Producer implement IAdmin, but I understand you don't want that, either, have an IAdminClient AdminHelper.CreateFrom(Producer), or a AdminClient(Producer), but AdminClient should not be disposable in this case, so i don't like it - as the handle implements IAdmin, it will be no problem)

Reply:
disposabiltiy is a good point that i didn't think of above.

Reply:
Other solution for consumer / producer: only expose producer / consumer api (Producer only expose consume), and have an Handle / Admin property which allow access on IHandle / IAdmin api (producer.Handle.GetMetadata, producer.Admin.CreateTopics) I don't know if I like this (great separation of api) or dislike it (harder to call basic method on producer/consumer)

Reply:
Yeah, that idea has come up before and been discarded.

Another point is the Dispose method doesn't necessarily need to dispose the SafeKafkaHandle. If an AdminClient was constructed using a Handle, the Dispose method could do nothing. Alternatively we could reference count and dispose when the last reference is disposed. 

(just some quick brainstorming - no clear opinion yet).

Reply:
I'm going off the idea of having any public methods on Handle (in part because of the usage you point out above). Maybe we have just Producer, Consumer, and AdminClient (and corresponding interfaces as you suggest).

Maybe typical usage is like:

```
using (var adminClient = new AdminClient(config))
{
    ...
}
```

or 

```
using (var producer = new Producer<string, string>(config, new StringSerializer(Encoding.UTF8), new StringSerializer(Encoding.UTF8)))
using (var adminClient = new AdminClient(producer.Handle))
{
    ...
}
```

but disposal of adminClient in the latter case actually does nothing. And maybe it's ok that if the producer gets destroyed, adminClient can no longer be used.

Reply:
the below is just brainstorming - thinking out load. thought it might help to write this down, but i don't think it's going anywhere.

It'd be kind of cool if this were possible:

```
using (var client = new KafkaClient(config))
{
    var producer = (IProducer<string, string>)client;
    var adminClient = (IAdminClient)client;
    var consumer = (IConsumer)client;
}
```

where explicit interface implementations are used (as you are suggesting). 

of course we can't convert between Producer and Consumer. Perhaps:

```
using (var client = new KafkaClient(config, ClientType.Producer))
{
    var producer = (IProducer<string, string>)client;
    var adminClient = (IAdminClient)client;
    var consumer = (IConsumer)client; // disallowed
}
```

and/or

```
using (var producer = KafkaClient.CreateProducer<string, string>(config, new StringSerializer(Encoding.UTF8), new StringSerializer(Encoding.UTF8)))
{
    var adminClient = (IAdminClient)producer; 
}
```

where CreateProducer returns IProducer<string,string> and is an instance of`KafkaClient` (type Producer).

This is quite a departure from where we are currently of course. I'm not sure it's better.

The Handle idea also isn't ideal, but i think it's fairly inoffensive.

With java, you can't reuse connections for the different client types. We could also do that (but it'd be a pity).

Reply:
Maybe we:

1. Have a new client type `AdminClient`. We don't allow this to be constructed in a way which reuses a librdkafka handle. It's a new client. Same as in Java. Same as what you have done in this PR (good job!)

2. Additionally, allow an explicit cast of `Producer` or `Consumer` to `IAdminClient` as kind of a backdoor way of re-using these connections for admin client tasks. Same as what you have in this PR (good job!). People have dissuaded me from using explicit casts here in the past, but i'm thinking it's a good idea. It's kind of an advanced usage.

3. Make `GetMetadata`, `DeleteTopics`, `ListGroups` only be available via the IAdminClient interface to make the `Producer` and `Consumer` APIs nice and clean.

4. Figure out some nice factoring to reduce code duplication (though not necessarily in this PR).